src/micras.cpp

```diff
@@ -34,14 +38,104 @@ void Micras::update() {
     this->elapsed_time = loop_stopwatch.elapsed_time_us() / 1e6F;
     loop_stopwatch.reset_us();
 
-    this->button_status = button.get_status();
-
+    this->button.update();
     this->buzzer.update();
     this->fan.update();
     this->imu->update();
     this->wall_sensors->update();
-    this->fsm.run();
+    this->fsm.update();
 
     while (loop_stopwatch.elapsed_time_us() < loop_time_us) { }
 }
+
+bool Micras::calibrate() {
+    switch (this->calibration_type) {
+        case CalibrationType::SIDE_WALLS:
+            this->wall_sensors->calibrate_left_wall();
+            this->wall_sensors->calibrate_right_wall();
+            this->calibration_type = CalibrationType::FRONT_WALL;
+            return false;
+
+        case CalibrationType::FRONT_WALL:
+            this->wall_sensors->calibrate_front_wall();
+            this->calibration_type = CalibrationType::SIDE_WALLS;
+            this->wall_sensors->turn_off();
+            return true;
+    }
+
+    return false;
+}
+
+bool Micras::run(float elapsed_time) {
+    this->odometry.update(elapsed_time);
+
+    const micras::nav::State& state = this->odometry.get_state();
+    core::Observation         observation{};
+
+    if (this->current_action->finished(this->action_pose.get())) {
+        if (this->finished) {
+            this->finished = false;
+            return true;
+        }
+
+        this->speed_controller.reset();
+        this->action_pose.reset_reference();
+
+        if (not this->action_queuer.empty()) {
+            this->current_action = this->action_queuer.pop();
+        } else {
+            const bool returning = (this->objective == core::Objective::RETURN);
+            const bool solving = (this->objective == core::Objective::SOLVE);
+
+            if (not solving) {
+                observation = this->wall_sensors->get_observation();
+                this->maze.update_walls(this->grid_pose, observation);
+            }
+
+            micras::nav::GridPose next_goal{};
+
+            if (solving or this->maze.finished(this->grid_pose.position, returning)) {
+                this->finished = true;
+                next_goal = this->grid_pose.turned_back().front();
```

Comment:
pq isso aqui?

Reply:
Isso é pq se eu dou push no action queuer de uma célula atrás do robô, ele coloca 3 actions na fila, uma de parar andando meia célula, uma de girar 180º e uma de andar meia célula, ai nesse caso ele termina de explorar e começa a executar essa ação de parar, mas quando termina ele vê que a flag finished ta setada e vai pro estado de wait, e quando termina de esperar, tem duas ações na fila já, uma de girar e outra de andar meia célula, que é o que ele precisa fazer pra começar a retornar pro começo.

Reply:
certo isso aqui pro caso do `this->maze.finished(this->grid_pose.position, returning))`, mas e o caso do `solving`? O solving ainda teria que ser dividio em acoes, mas do jeito que ta eu entendo que no final da primeira acao ele já entraria nesse if

Reply:
Pro solving ele vai executar a ação de stop e vai voltar pro estado de idle, ai a fila é resetada pra ter só a ação inicial.